Document config system

leotrs opened the pull request:
## List of Changes
- Created a new section inside Reference
- Documented all private modules classes methods within the config sub package
- renamed the config subpackage from manim/config to manim/_config, and the logger module from manim/_config/logger.py to manim/_config/logger_utils.py. This is made so that there are no longer two objects with the same name (the global `config` object and the `config` package, and the global `logger` object and the `logger` module. This was screwing up with the docs build.)

## Testing Status
Tests and docs build. Doctests do not pass. @behackl could you help me here? I'm not familiar with them and I tried fixing them, with no success.

## Further comments
I am marking this as draft for now so that others can chime in as I keep working on it. My next step is to write a tutorial document on configuring manim.

leotrs commented:
Ignore the extraneous commits, I had to merge from master twice.

behackl commented:
When doctests are executed, a new Python intrrpreter is spawned to execute the lines. I think adding `from manim import config` as the first line should resolve it.

leotrs commented:
Ok I believe I fixed the main issues. However, there are a couple of doctests whose output depends on the system. Pathlib will return a PosixPath on linux but a WindowsPath on windows. Haven't found a way of making these pass on both systems.

leotrs commented:
Alright this is ready for review @ManimCommunity/docs. Still can't fix the Windows problem.

Minor changes:
+ I created a new section in the [Reference index](https://manimce--659.org.readthedocs.build/en/659/reference.html)
+ Changed the TOC of the [Tutorials index](https://manimce--659.org.readthedocs.build/en/659/tutorials.html)
+ Updated the [changelog](https://manimce--659.org.readthedocs.build/en/659/changelog.html)

Major changes:
+ [Configuration tutorial](https://manimce--659.org.readthedocs.build/en/659/tutorials/configuration.html)
+ Module docstrings [here](https://manimce--659.org.readthedocs.build/en/659/reference/manim._config.html#module-manim._config) and [here](https://manimce--659.org.readthedocs.build/en/659/reference/manim._config.utils.html#module-manim._config.utils)
+ [ManimConfig class docstring](https://manimce--659.org.readthedocs.build/en/659/reference/manim._config.utils.ManimConfig.html#manim._config.utils.ManimConfig)

behackl commented on manim/_config/utils.py:
```diff
@@ -743,10 +1037,119 @@ def use_js_renderer(self, val):
+            '{media_dir}/Tex'
+            >>> config.media_dir
+            './media'
+            >>> config.get_dir("tex_dir")
+            PosixPath('media/Tex')
```
Here is one possible way to make this test OS-agnostic:

```suggestion
            >>> config.get_dir("tex_dir").as_posix()
            'media/Tex'
```

leotrs replied:
Ah good one!

leotrs commented:
Thanks @behackl it worked!

Now for sure ready for review.

leotrs commented:
@ManimCommunity/docs

cobordism commented on docs/source/tutorials/configuration.rst:
```diff
@@ -135,76 +101,88 @@ open the file after it is rendered.
-.. warning:: The config file **must** be named :code:`manim.cfg`. Currently,
-             manim does not support config files with any other name.
+.. warning:: The config file **must** be named ``manim.cfg``. Currently, manim
+             does not support config files with any other name.
```
is this even true? What about --config_file ?

leotrs replied:
This is true in the sense that the user and folder files can only have that name. If you use --config_file, the file can have any name. I agree this is confusing. Open to suggestions on how to implement it. (The problem is that the user and folder files are loaded and parsed way before the CLI arguments are parsed. So there's currently no way of changing that without a *major* refactor...)

kolibril13 commented on manim/mobject/types/image_mobject.py:
```diff
@@ -146,7 +179,7 @@ class ImageMobjectFromCamera(AbstractImageMobject):
         self.camera = camera
         self.pixel_array = self.camera.pixel_array
-        AbstractImageMobject.__init__(self, **kwargs)
+        AbstractImageMobject.__init__(self, scale_to_resolution=False, **kwargs)
```
interesting.
These are changes I did in a previous pr.
Why are they occurring here?

kolibril13 commented on tests/helpers/path_utils.py:
```diff
@@ -0,0 +1,5 @@
+
+
+def get_project_root() -> Path:
+    return Path(__file__).parent.parent.parent
```
also from another pr

leotrs commented:
I believe they come from the time I merged from mssteem the diff is shown
with respect to the commit at which I started the branch,  not to the
current master.

On Thu, Nov 5, 2020, 4:13 AM kolibril13 <notifications@github.com> wrote:

> *@kolibril13* commented on this pull request.
> ------------------------------
>
> In tests/helpers/path_utils.py
> <https://github.com/ManimCommunity/manim/pull/659#discussion_r517897097>:
>
> > @@ -0,0 +1,5 @@
> +from pathlib import Path
> +
> +
> +def get_project_root() -> Path:
> +    return Path(__file__).parent.parent.parent
>
> also from another pr
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ManimCommunity/manim/pull/659#pullrequestreview-524059246>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAILYAHMQBJX2WYSI6LLWUTSOJUDVANCNFSM4TGIL5CA>
> .
>

kolibril13 commented:
ok, I see. But then I am confused that the difference shows that tests/helpers/path_utils.py. As this file already exists at master, it should not be shown as added. Strange... But merging master into this commit would solve that, right?

leotrs commented:
I'm not sure to be honest. Also, I'm much more interested in your thoughts
about the actual documentation I wrote than in having the commits organized
:)

I know our commit history is already messy as it is, but the more we delay
review of this PR, the more times I will have to merge from master and the
messier it will become!

On Thu, Nov 5, 2020 at 8:12 AM kolibril13 <notifications@github.com> wrote:

> ok, I see. But then I am confused that the difference shows that
> tests/helpers/path_utils.py. As this file already exists at master, it
> should not be shown as added. Strange... But merging master into this
> commit would solve that, right?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ManimCommunity/manim/pull/659#issuecomment-722369137>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAILYAHIRKI76A7URHCZMNTSOKQD3ANCNFSM4TGIL5CA>
> .
>


-- 
www.leotrs.com | leo@leotrs.com
PhD candidate at the Network Science Institute, Northeastern University

kolibril13 commented:
> I'm not sure to be honest. Also, I'm much more interested in your thoughts about the actual documentation I wrote than in having the commits organized :) I know our commit history is already messy as it is, but the more we delay review of this PR, the more times I will have to merge from master and the messier it will become!
> […](#)
> On Thu, Nov 5, 2020 at 8:12 AM kolibril13 ***@***.***> wrote: ok, I see. But then I am confused that the difference shows that tests/helpers/path_utils.py. As this file already exists at master, it should not be shown as added. Strange... But merging master into this commit would solve that, right? — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub <[#659 (comment)](https://github.com/ManimCommunity/manim/pull/659#issuecomment-722369137)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAILYAHIRKI76A7URHCZMNTSOKQD3ANCNFSM4TGIL5CA> .
> -- [www.leotrs.com](http://www.leotrs.com) | leo@leotrs.com PhD candidate at the Network Science Institute, Northeastern University

Good point!
I asked only to know what actually is code that needs to be reviewed and what code is already in master, so that I know what to review.
But I will try figure that out myself :)

leotrs commented:
See [here](https://github.com/ManimCommunity/manim/pull/659#issuecomment-720144391)

kolibril13 commented on docs/source/tutorials/configuration.rst:
```diff
@@ -1,125 +1,91 @@
 
-Another frequent flag is :code:`-p`, which makes manim show the rendered video
-right after it's done rendering.
+   >>> Camera({}).background_color
```
Is the empty dict necessary here?

leotrs replied:
Yes.

kolibril13 commented:
Just saw that in https://manimce--659.org.readthedocs.build/en/659/tutorials/configuration.html
we still have this:
![image](https://user-images.githubusercontent.com/44469195/98252641-77202a80-1f7a-11eb-9c55-9cda0eca7962.png)
Moreover, I think a short diagram would be nice to show that there are three options to set configurations: 1. CLI flags, 2. config file, 3. config.sssommething..

kolibril13 commented:
https://manimce--659.org.readthedocs.build/en/659/reference/manim._config.html#module-manim._config
![image](https://user-images.githubusercontent.com/44469195/98253194-04fc1580-1f7b-11eb-9bc0-c863b52324eb.png)
Why is there this "True" parameter?

leotrs commented:
> Why is there this "True" parameter?

Good call, I fixed that and turned it into a doctest.

leotrs commented:
The background opacity flag was removed from a PR after this one. So when this is merged that should work itself out.

behackl commented on manim/_config/main_utils.py:
```diff
@@ -1,10 +1,4 @@
-Functions called from __main__.py to interact with the config.
-
-"""
+"""Utilities called from __main__.py to interact with the config."""
```
```suggestion
"""Utilities called from ``__main__.py`` to interact with the config."""
```

behackl commented on manim/_config/utils.py:
```diff
@@ -19,10 +21,43 @@
 
 
 def config_file_paths():
+    """The paths where .cfg files will be searched for.
```
```suggestion
    """The paths where ``.cfg`` files will be searched for.
```

behackl commented on manim/_config/utils.py:
```diff
@@ -19,10 +21,43 @@
 def config_file_paths():
+    """The paths where .cfg files will be searched for.
+
+    When manim is first imported, it processes any .cfg files it finds.  This
```
```suggestion
    When manim is first imported, it processes any ``.cfg`` files it finds.  This
```

behackl commented on manim/_config/utils.py:
```diff
@@ -19,10 +21,43 @@
+    Returns
+    -------
+    List[:class:`Path`]
+        List of paths which may contain .cfg files, in ascending order of
```
```suggestion
        List of paths which may contain ``.cfg`` files, in ascending order of
```

behackl commented on manim/_config/utils.py:
```diff
@@ -1,8 +1,10 @@
+contains all configuration options, including frame geometry (e.g. frame
+height/width, frame rate), output (e.g. directories, logging), styling
+(e.g. background color, transparency), and general behavior (e.g. writing a
+movie vs writing a single frame).
```
I think it makes sense to add a reference to the configuration tutorial here as well.

```suggestion
movie vs writing a single frame).

See :doc:`tutorials/configuration` for an introduction to Manim's configuration system.
```

behackl commented on manim/_config/utils.py:
```diff
@@ -33,20 +68,31 @@ def config_file_paths():
 
 def make_config_parser(custom_file=None):
-    """Make a ConfigParser object and load the .cfg files.
+    """Make a :class:`ConfigParser` object and load any .cfg files.
```
```suggestion
    """Make a :class:`ConfigParser` object and load any ``.cfg`` files.
```

behackl commented on manim/_config/utils.py:
```diff
@@ -93,6 +139,108 @@ def determine_quality(args):
+       [CLI]
+       background_color = WHITE
+
+    In order to have this .cfg file apply to a manim scene, it needs to be
```
```suggestion
    In order to have this ``.cfg`` file apply to a manim scene, it needs to be
```

behackl commented on manim/_config/utils.py:
```diff
@@ -93,6 +139,108 @@ def determine_quality(args):
+
+        manim scene.py
+
+    the background of the scene will be set to WHITE.  This applies regardless
```
```suggestion
    the background of the scene will be set to ``WHITE``.  This applies regardless
```

behackl commented on manim/_config/utils.py:
```diff
@@ -93,6 +139,108 @@ def determine_quality(args):
+    the background of the scene will be set to WHITE.  This applies regardless
+    of where the manim command is invoked from.
+
+    Command line arguments override .cfg files.  In the previous example,
```
```suggestion
    Command line arguments override ``.cfg`` files.  In the previous example,
```

behackl commented on manim/_config/utils.py:
```diff
@@ -238,6 +436,55 @@ def _set_pos_number(self, key, val, allow_inf):
+        Parameters
+        ----------
+        parser : :class:`ConfigParser`
+            An object reflecting the contents of one or many .cfg files.  In
```
```suggestion
            An object reflecting the contents of one or many ``.cfg`` files.  In
```

behackl commented on manim/_config/utils.py:
```diff
@@ -238,6 +436,55 @@ def _set_pos_number(self, key, val, allow_inf):
+        parser : :class:`ConfigParser`
+            An object reflecting the contents of one or many .cfg files.  In
+            particular, it may reflect the contents of mulitple files that have
+            been parsed in a cascading fasion.
```
```suggestion
            been parsed in a cascading fashion.
```

behackl commented on manim/_config/utils.py:
```diff
@@ -238,6 +436,55 @@ def _set_pos_number(self, key, val, allow_inf):
+
+        See Also
+        --------
+        :func:`make_config_parser`, :meth:`~ManimConfig.digest_file`,
+        :meth:`~ManimConfig.digest_args`,
```
```suggestion
        :func:`make_config_parser`, :meth:`~.ManimConfig.digest_file`,
        :meth:`~.ManimConfig.digest_args`,
```

behackl commented on manim/_config/utils.py:
```diff
@@ -238,6 +436,55 @@ def _set_pos_number(self, key, val, allow_inf):
+
+        Notes
+        -----
+        If there are multiple .cfg files to process, it is always more
```
```suggestion
        If there are multiple ``.cfg`` files to process, it is always more
```

behackl commented on manim/_config/utils.py:
```diff
@@ -238,6 +436,55 @@ def _set_pos_number(self, key, val, allow_inf):
+        If there are multiple .cfg files to process, it is always more
+        efficient to parse them into a single :class:`ConfigParser` object
+        first, and then call this function once (instead of calling
+        :meth:`~ManimConfig.digest_file` multiple times).
```
```suggestion
        :meth:`~.ManimConfig.digest_file` multiple times).
```

behackl commented on manim/_config/utils.py:
```diff
@@ -312,6 +559,30 @@ def digest_parser(self, parser):
+
+        See Also
+        --------
+        :func:`.main_utils.parse_args()`, :meth:`~ManimConfig.digest_parser`,
+        :meth:`~ManimConfig.digest_file`
```
```suggestion
        :func:`.main_utils.parse_args()`, :meth:`~.ManimConfig.digest_parser`,
        :meth:`~.ManimConfig.digest_file`
```

behackl commented on manim/_config/utils.py:
```diff
@@ -312,6 +559,30 @@ def digest_parser(self, parser):
+
+        Notes
+        -----
+        If ``args.config_file`` is non-null, ManimConfig tries to digest the
```
non-null does not sound very pythonic. "not ``None``"? Or maybe even "If ``args.config_file`` is set, ...?

leotrs replied:
```suggestion
        If ``args.config_file`` is a non-empty string, ``ManimConfig`` tries to digest the
```

behackl commented on manim/_config/utils.py:
```diff
@@ -409,41 +680,68 @@ def digest_args(self, args):
-        pass
-
     def digest_file(self, filename):
+        """Process the config options present in a .cfg file.
```
```suggestion
        """Process the config options present in a ``.cfg`` file.
```

behackl commented on manim/_config/utils.py:
```diff
@@ -409,41 +680,68 @@ def digest_args(self, args):
     def digest_file(self, filename):
+        """Process the config options present in a .cfg file.
+
+        This method processes a single .cfg file, whereas
```
```suggestion
        This method processes a single ``.cfg`` file, whereas
```

behackl commented on manim/_config/utils.py:
```diff
@@ -409,41 +680,68 @@ def digest_args(self, args):
+
+        This method processes a single .cfg file, whereas
+        :meth:`~ManimConfig.digest_parser` can process arbitrary parsers, built
+        perhaps from multiple .cfg files.
```
```suggestion
        perhaps from multiple ``.cfg`` files.
```

behackl commented on manim/_config/utils.py:
```diff
@@ -409,41 +680,68 @@ def digest_args(self, args):
+        Parameters
+        ----------
+        filename : :class:`str`
+            Path to the .cfg file.
```
```suggestion
            Path to the ``.cfg`` file.
```

behackl commented on manim/_config/utils.py:
```diff
@@ -409,41 +680,68 @@ def digest_args(self, args):
+
+        Notes
+        -----
+        If there are multiple .cfg files to process, it is always more
```
```suggestion
        If there are multiple ``.cfg`` files to process, it is always more
```

behackl commented on manim/_config/utils.py:
```diff
@@ -743,10 +1037,119 @@ def use_js_renderer(self, val):
+        ----------
+        key : :class:`str`
+            The config option to be resolved.  Must be an option ending in
+            '_dir', for example 'media_dir' or 'video_dir'.
```
```suggestion
            ``'_dir'``, for example ``'media_dir'`` or ``'video_dir'``.
```

behackl commented on manim/_config/utils.py:
```diff
@@ -743,10 +1037,119 @@ def use_js_renderer(self, val):
+        -------
+        :class:`pathlib.Path`
+            Path to the requested directory.  If the path resolves to the empty
+            string, return None instead.
```
```suggestion
            string, return ``None`` instead.
```

behackl commented on manim/_config/utils.py:
```diff
@@ -743,10 +1037,119 @@ def use_js_renderer(self, val):
+        Examples
+        --------
+
+        The value of ``config.tex_dir`` is ``{media_dir}/Tex`` by default,
```
```suggestion
        The value of ``config.tex_dir`` is ``'{media_dir}/Tex'`` by default,
```

behackl reviewed (changes requested):
This looks excellent, thank you very much for your effort documenting the config system -- it is great to have a new resource in our docs allowing us to easily answer questions like "how do I change the background color?". :-)

I've made a bunch of (non-blocking) suggestions for adding code literals, one or two typos -- just superficial stuff; give it a short look, we can immediately merge this after you have done so.

kolibril13 commented:
I am a fan of graphical mnemonics, so I created this here:
![Copy of nice_font_manim_under_the_hood (1)](https://user-images.githubusercontent.com/44469195/98421001-af159380-2088-11eb-961c-2dba43d35f99.png)
What do you think?
you can edit it here:
https://drive.google.com/file/d/1WYVKKoRbXrumHEcyQKQ9s1yCnBvfU2Ui/view?usp=sharing

kolibril13 commented:
And I just saw we still have 

# --background_opacity
background_opacity = 1

in default.cfg.
Can we remove this?

kolibril13 commented:
And one more thought, maybe for a future pr,
at the moment, it is possible to run scenes like this:
```python
from manim import *
from manim import config
import matplotlib.pyplot as plt
class Example(Scene):
    def construct(self):
        circle = Circle()
        self.add((circle))

config["write_to_movie"] = False
config["save_last_frame"] = True
scene = Example()
scene.render()
data = scene.renderer.get_frame()
plt.imshow(data)
plt.show()
```
I think this is quite cool, and could maybe also be included somewhere in the docs.

leotrs commented:
> I am a fan of graphical mnemonics, so I created this here:
> What do you think?

This looks good to me, and it would be a fine addition. However, right now I'm not sure what the diagram is conveying. The arrows seem to follow the ways of setting the config in increasing order of precedence. But since the structure of the diagram is linear, just a simple linked list would do, no? If so, I don't see how this diagram, in its current iteration, is better than the itemized list currently in the "Order of operations" section in the new tutorial.

What I would *love* to see instead is a [flowchart](https://en.wikipedia.org/wiki/Flowchart) that users can follow to decide which way of manipulating the config is the best, with forking paths and all. For example this could include a decision diamond with the question "does the config option you want to manipulate have a CLI flag?" and then presenting different choices depending on the answer yes/no.

leotrs commented:
> And one more thought, maybe for a future pr,
> at the moment, it is possible to run scenes like this:
> I think this is quite cool, and could maybe also be included somewhere in the docs.

I think this is more about the `render()` method of Scene rather than the config system. So I'd put it in that class.

leotrs commented:
> And I just saw we still have
> 
> # --background_opacity
> background_opacity = 1
> 
> in default.cfg.
> Can we remove this?

No. There still is a config option for background opacity, just not a CLI flag anymore. The options in default.cfg serve as default values for all config options, not only the CLI flags.
